### ISSUE TITLE
Nutrihuman

### ISSUE THREAD

**Issue by Tshaynyb** [Tshaynyb/Eddy#1]
import streamlit as st
from PIL import Image
import numpy as np
import tensorflow as tf
from tensorflow.keras.applications.efficientnet import preprocess_input, decode_predictions
from tensorflow.keras.preprocessing import image
import sqlite3
from datetime import datetime
import os
import uuid

# Load a pre-trained food-specific model (EfficientNet trained on ImageNet)
model = tf.keras.applications.EfficientNetB0(weights='imagenet') # Replace with a Food-101 model if available

# Function to analyze meal images
def analyze_meal(image_path):
try:
# Load and preprocess the image
img = image.load_img(image_path, target_size=(224, 224))
img_array = image.img_to_array(img)
img_array = np.expand_dims(img_array, axis=0)
img_array = preprocess_input(img_array)

# Predict ingredients using the model
predictions = model.predict(img_array)
top_preds = decode_predictions(predictions, top=5)[0]

# Extract food-related items from predictions
detected_ingredients = [pred[1] for pred in top_preds] # pred[1] contains the human-readable label

return detected_ingredients
except Exception as e:
st.error(f"Error analyzing meal: {e}")
return []

# User profile class
class UserProfile:
def __init__(self, name, dob, height, weight):
[self.name](http://self.name/) = name
self.dob = dob
self.height = height # in inches
self.weight = weight # in pounds
self.age = self.calculate_age()
self.bmi = self.calculate_bmi()
self.goal = None # e.g., "lose_weight", "improve_energy", "support_digestion"
self.health_categories = []

def calculate_age(self):
try:
today = datetime.today()
birth_date = datetime.strptime(self.dob, "%m/%d/%Y")
age = today.year - birth_date.year
if (today.month, today.day) < (birth_date.month, birth_date.day):
age -= 1
return age
except Exception as e:
st.error(f"Error calculating age: {e}")
return 0

def calculate_bmi(self):
try:
height_meters = self.height * 0.0254 # Convert height to meters
weight_kg = self.weight * 0.453592 # Convert weight to kilograms
bmi = weight_kg / (height_meters ** 2)
return round(bmi, 1)
except Exception as e:
st.error(f"Error calculating BMI: {e}")
return 0

def set_goal(self, goal):
self.goal = goal

def add_health_category(self, category):
self.health_categories.append(category)

def daily_calorie_needs(self):
try:
# Basal Metabolic Rate (BMR) calculation using Mifflin-St Jeor equation
bmr = 10 * (self.weight * 0.453592) + 6.25 * (self.height * 2.54) - 5 * self.age + 5
if self.goal == "lose_weight":
return int(bmr * 1.2) - 500 # Subtract 500 calories for weight loss
elif self.goal == "gain_weight":
return int(bmr * 1.2) + 500 # Add 500 calories for weight gain
else:
return int(bmr * 1.2) # Default maintenance calories
except Exception as e:
st.error(f"Error calculating calorie needs: {e}")
return 2000

# Meal and juice plans
class MealPlans:
def __init__(self, user_profile):
self.user_profile = user_profile

def generate_weekly_meal_plan(self):
meal_plan = {}
for category in self.user_profile.health_categories:
if category == "performance_energy":
meal_plan.update({
"Monday": "Quinoa bowl with hemp seeds, walnuts, and kale.",
"Tuesday": "Chickpea salad with dandelion greens and watercress.",
"Wednesday": "Amaranth porridge with burro bananas and berries.",
"Thursday": "Fonio stir-fry with okra and soursop juice.",
"Friday": "Sea moss smoothie with dates and figs.",
"Saturday": "Walnut milk with burro bananas and chia seeds.",
"Sunday": "Grilled tofu with kale and quinoa."
})
elif category == "digestive_health":
meal_plan.update({
"Monday": "Teff porridge with figs and tamarind juice.",
"Tuesday": "Wild rice salad with nopales and okra.",
"Wednesday": "Chickpea curry with dandelion greens.",
"Thursday": "Sea moss smoothie with papaya and aloe vera water.",
"Friday": "Fonio with steamed vegetables and tamarind juice.",
"Saturday": "Fermented vegetables with hemp seeds.",
"Sunday": "Quinoa bowl with kale and figs."
})
# Add more categories as needed...
return meal_plan

def generate_weekly_juice_plan(self):
juice_plan = {}
for category in self.user_profile.health_categories:
if category == "performance_energy":
juice_plan.update({
"Monday": "Burro banana smoothie.",
"Tuesday": "Soursop juice.",
"Wednesday": "Walnut milk.",
"Thursday": "Date and fig juice.",
"Friday": "Berry blast smoothie.",
"Saturday": "Green juice with kale and cucumber.",
"Sunday": "Hemp seed milk."
})
elif category == "digestive_health":
juice_plan.update({
"Monday": "Papaya and tamarind juice.",
"Tuesday": "Aloe vera water.",
"Wednesday": "Sea moss smoothie.",
"Thursday": "Ginger and tamarind shot.",
"Friday": "Cucumber and mint juice.",
"Saturday": "Apple and celery juice.",
"Sunday": "Kale and pear juice."
})
# Add more categories as needed...
return juice_plan

# SQLite database integration
def init_db():
conn = sqlite3.connect('nutrihuman.db', check_same_thread=False)
cursor = conn.cursor()

# Create tables for users and meal plans
cursor.execute('''
CREATE TABLE IF NOT EXISTS users (
id INTEGER PRIMARY KEY AUTOINCREMENT,
name TEXT NOT NULL,
dob TEXT NOT NULL,
height REAL NOT NULL,
weight REAL NOT NULL,
goal TEXT NOT NULL,
health_categories TEXT NOT NULL
)
''')

cursor.execute('''
CREATE TABLE IF NOT EXISTS meal_plans (
id INTEGER PRIMARY KEY AUTOINCREMENT,
user_id INTEGER NOT NULL,
day TEXT NOT NULL,
meal TEXT NOT NULL,
FOREIGN KEY (user_id) REFERENCES users (id)
)
''')

conn.commit()
return conn

# Function to add a user
def add_user(conn, name, dob, height, weight, goal, health_categories):
cursor = conn.cursor()
cursor.execute('''
INSERT INTO users (name, dob, height, weight, goal, health_categories)
VALUES (?, ?, ?, ?, ?, ?)
''', (name, dob, height, weight, goal, ",".join(health_categories)))
conn.commit()
return cursor.lastrowid # Return the user ID

# Function to add a meal plan
def add_meal_plan(conn, user_id, day, meal):
cursor = conn.cursor()
cursor.execute('''
INSERT INTO meal_plans (user_id, day, meal)
VALUES (?, ?, ?)
''', (user_id, day, meal))
conn.commit()

# Function to get a user's meal plan
def get_meal_plan(conn, user_id):
cursor = conn.cursor()
cursor.execute('''
SELECT day, meal FROM meal_plans WHERE user_id = ?
''', (user_id,))
return cursor.fetchall()

# Streamlit app
def main():
st.title("NutriHuman - Meal Image Analysis")
st.write("Upload an image of your meal to analyze its ingredients.")

# Initialize database
conn = init_db()

# Session state management
if "user_id" not in st.session_state:
st.session_state.user_id = None

# User profile creation
st.sidebar.header("User Profile")
name = st.sidebar.text_input("Enter your name:")
dob = st.sidebar.text_input("Enter your date of birth (MM/DD/YYYY):")
height = st.sidebar.number_input("Enter your height (in inches):", min_value=0.0)
weight = st.sidebar.number_input("Enter your weight (in pounds):", min_value=0.0)
goal = st.sidebar.selectbox("What is your primary goal?", ["lose_weight", "improve_energy", "support_digestion"])
health_categories = st.sidebar.multiselect("Select health categories:", ["performance_energy", "digestive_health"])

if st.sidebar.button("Create Profile"):
if name and dob and height > 0 and weight > 0:
user_profile = UserProfile(name, dob, height, weight)
user_profile.set_goal(goal)
for category in health_categories:
user_profile.add_health_category(category)
st.session_state.user_id = add_user(conn, name, dob, height, weight, goal, health_categories)
st.sidebar.success("Profile created successfully!")
else:
st.sidebar.error("Please fill in all fields correctly.")

# Meal image analysis
uploaded_file = st.file_uploader("Choose an image...", type=["jpg", "jpeg", "png"])
if uploaded_file is not None:
# Generate a unique filename to prevent overwriting
unique_filename = str(uuid.uuid4()) + ".jpg"
image_path = os.path.join("uploads", unique_filename)
os.makedirs("uploads", exist_ok=True)

# Save the uploaded image
with open(image_path, "wb") as f:
f.write(uploaded_file.getbuffer())

# Display the uploaded image
image = Image.open(image_path)
st.image(image, caption='Uploaded Meal', use_column_width=True)
st.write("Analyzing...")

# Analyze the meal
ingredients = analyze_meal(image_path)
st.write("Identified Ingredients:")
st.write(ingredients)

# Display meal plans
if st.button("Generate Meal Plan"):
if st.session_state.user_id:
meal_plans = get_meal_plan(conn, st.session_state.user_id)
st.write("--- Weekly Meal Plan ---")
for day, meal in meal_plans:
st.write(f"{day}: {meal}")
else:
st.error("Please create a profile first.")

# Run the app
if __name__ == "__main__":
main()
